### PR TITLE
Course list/bl 97

### PR DESCRIPTION
**Pull request by pmAllport**
* Upgraded from bloc_test 7.0.0 -> 8.0.0
* Upgraded from flutter_bloc 6.0.0 -> 7.0.0
* Upgraded from bloc 6.0.0 -> 7.0.0
* Added mocktail 0.1.1 to dependencies (used by bloc-test)
* Fixed breaking errors of tests when upgrading from bloc_test 7.0.0 -> 8.0.0 (ref : https://pub.dev/packages/bloc_test/changelog) 
* Fixed "onError"-funtion in lib/blocs/simple_bloc_observer.dart

### CODE REVIEW

**Comment by pmAllport** [comment]
Need to finish development before we push to master

File: lib/blocs/simple_bloc_observer.dart [hunk 1]
```diff
@@ -14,8 +14,8 @@ class SimpleBlocObserver extends BlocObserver {
   }
 
   @override
-  void onError(Cubit cubit, Object error, StackTrace stackTrace) {
-    print('onError $error');
-    super.onError(cubit, error, stackTrace);
+  void onError(BlocBase bloc, Object error, StackTrace stackTrace) {
```
Comment (amundlrohne) [inline lib/blocs/simple_bloc_observer.dart:17]: Is this for better stacktraces?

File: lib/views/course/alternative.dart [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+              isQuiz: true));
+        },
+        child: Container(
+          constraints: BoxConstraints(
+              maxWidth: MediaQuery.of(context).size.width / 2.5,
+              maxHeight: MediaQuery.of(context).size.width / 2.5),
+          alignment: FractionalOffset.center,
+          decoration: widget.isPressed
+              ? BoxDecoration(
+                  borderRadius: BorderRadius.circular(5),
```
Comment (amundlrohne) [inline lib/views/course/alternative.dart:45]: Only border radius if pressed?

File: lib/views/course/alternative.dart [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  @override
+  _AlternativeState createState() => _AlternativeState();
+}
+
+class _AlternativeState extends State<Alternative> {
+  @override
+  Widget build(BuildContext context) {
+    return GestureDetector(
+        onTap: () {
+          debugPrint('triggerd');
```
Comment (amundlrohne) [inline lib/views/course/alternative.dart:31]: remove

File: lib/views/course/alternative_container.dart [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+          onTap: () => {
+                setState(() => alts = alts.sublist(0, 1)),
+                debugPrint('CLICKCLICK: ' + alts.toString()),
+
+                /* BlocProvider.of<CourseDetailBloc>(context).add(
+                    CourseDetailRequested(
+                        course: widget.course,
+                        isAnswer: true,
+                        answerId: 1,
+                        isQuiz: true)) */
```
Comment (amundlrohne) [inline lib/views/course/alternative_container.dart:70]: remove

File: lib/views/course/alternative_container.dart [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+//             crossAxisCount: 2,
+//             childAspectRatio: 1,
+//             crossAxisSpacing: 20,
+//             mainAxisSpacing: 20,
+//             children: List.generate(list.length, (index) {
+//               return Alter.Alternative(
+//                   index, list[index]['name'], list[index]['correct'], null);
+//             })));
+//   }
+// }
```
Comment (amundlrohne) [inline lib/views/course/alternative_container.dart:98]: remove

File: lib/views/course/course.dart [hunk 6]
```diff
@@ -1,54 +1,104 @@
...
   @override
   Widget build(BuildContext context) {
-    return Query(
-        options: QueryOptions(document: gql(getCoursesQuery)),
-        builder: (QueryResult result,
-            {VoidCallback refetch, FetchMore fetchMore}) {
+    return BlocBuilder<CourseDetailBloc, CourseDetailState>(
+      // ignore: missing_return
+      builder: (context, state) {
+        //log(state.toString());
```
Comment (amundlrohne) [inline lib/views/course/course.dart:41]: You guessed it

File: lib/views/course/course.dart [hunk 7]
```diff
@@ -59,6 +109,15 @@ class _CardContainerViewState extends State<CardContainerView> {
   @override
   void initState() {
     _controller = ScrollController();
+    _controller.addListener(() {
+      if (indx != (_controller.offset / context.size.width).round()) {
+        setState(() {
+          indx = (_controller.offset / context.size.width).roundToDouble();
+        });
+      }
+      print(indx);
```
Comment (amundlrohne) [inline lib/views/course/course.dart:118]: remove

File: test/utils/bloc/mock_bloc.dart [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+  Widget child,
+  Bloc bloc,
+}) {
+  return MaterialApp(
+    home: BlocProvider(
+      create: (context) => bloc,
+      child: child,
+    ),
+  );
+}
```
Comment (amundlrohne) [inline test/utils/bloc/mock_bloc.dart:15]: Where is this used?

File: lib/blocs/simple_bloc_observer.dart (hunk 1 above)
Comment (pmAllport) [reply lib/blocs/simple_bloc_observer.dart:17]: Should be. Its the one used in the bloc-documentation for BlocObservers : https://pub.dev/documentation/bloc/latest/

**Review by amundlrohne — COMMENTED** [review]
Implementation looks good. That being said, there are some things that should be cleaned up. Mostly commented out code, and debug prints. 

Also, will routing from course_list to course_detail be in a seperate PR?

File: lib/blocs/simple_bloc_observer.dart (hunk 1 above)
Comment (pmAllport) [reply lib/blocs/simple_bloc_observer.dart:17]: It is also worth mentioning, that this crashes on the newest version of bloc_test (8.0.0). I'm guessing its due to bloc_test implementing a "MockCubit", which means that that function no longer can cast to Bloc, as they are now two seperate functions.

File: lib/views/course/alternative.dart (hunk 3 above)
Comment (amundlrohne) [reply lib/views/course/alternative.dart:31]: The debugPrint that is

**Comment by pmAllport** [comment]
A lot of spread out comments here.

-The obvious dev-code will be removed. This is being worked on as we speak
-The tests have not yet been implemented. "CreateBlocProviderForTesting" is legacy and will be removed.
- The routing is also being worked on. Coming in its own PR.

File: lib/blocs/simple_bloc_observer.dart (hunk 1 above)
Comment (amundlrohne) [reply lib/blocs/simple_bloc_observer.dart:17]: > A lot of spread out comments here.
> 
> -The obvious dev-code will be removed. This is being worked on as we speak
> -The tests have not yet been implemented. "CreateBlocProviderForTesting" is legacy and will be removed.
> 
> * The routing is also being worked on. Coming in its own PR.

Awesome, will approve it. Merge when you feel ready! :)

**Review by amundlrohne — APPROVED** [review]
Approved!

**Comment by pmAllport** [comment]
* Pulled from master and resolved mergeconflicts
* Added some courseDetail tests (Event, State and Bloc).
* Still missing Course tests
* Fixed layout in slides
* Added button to end of slides to lead to "quiz"

**Comment by pmAllport** [comment]
Fixed flutter analyze issues

File: lib/blocs/course_detail/course_detail_bloc.dart [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        }
+      } catch (e, stackTrace) {
+        log(e.toString());
+        log(stackTrace.toString());
+
+        yield Failed(err: 'Error, bad request');
+        return;
+      }
+    }
+    yield Failed(err: 'Error, bad request');
```
Comment (amundlrohne) [inline lib/blocs/course_detail/course_detail_bloc.dart:88]: Good job with these Failed states. Would be nice with some clearer messages though.

File: lib/views/course/alternative_container.dart [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+                        alts[index]['correct'],
+                        alts[index]['image'],
+                        clicked));
+          }),
+        ),
+      )),
+      Container(
+          alignment: Alignment.topCenter,
+          padding: EdgeInsets.symmetric(vertical: 20),
+          child: Text(validation, style: TextStyle(color: Colors.teal))),
```
Comment (amundlrohne) [inline lib/views/course/alternative_container.dart:69]: Could be improved by retrieving textstyle from themes

**Review by amundlrohne — APPROVED** [review]
Looks good to me! A tad big, but all of the improvements in this PR should prevent this in the future. 👍

**Pull request merged**